src/routes_admin.py

```diff
@@ -0,0 +1,237 @@
+            return util.success(f"Successfully removed '{netid}'")
+
+        @blueprint.route("/staff/course/<cid>/<aid>/", methods=["POST"])
+        @auth.require_auth
+        @auth.require_admin_status
+        def staff_add_assignment(netid, cid, aid):
+            missing = util.check_missing_fields(request.form,
+                                                *["max_runs", "quota", "start", "end", "config", "visibility"])
+            if missing:
+                return util.error("Missing fields (%s)." % (", ".join(missing)))
+
+            if not util.valid_id(aid):
+                return util.error("Invalid Assignment ID. Allowed characters: a-z A-Z _ - .")
+
+            new_assignment = db.get_assignment(cid, aid)
+            if new_assignment:
+                return util.error("Assignment ID already exists.")
+
+            try:
+                max_runs = int(request.form["max_runs"])
+                if max_runs < 1:
+                    return util.error("Max Runs must be a positive integer.")
+            except ValueError:
+                return util.error("Max Runs must be a positive integer.")
+
+            quota = request.form["quota"]
+            if not db.Quota.is_valid(quota):
+                return util.error("Quota Type is invalid.")
+
+            start = util.parse_form_datetime(request.form["start"]).timestamp()
+            end = util.parse_form_datetime(request.form["end"]).timestamp()
+            if start is None or end is None:
+                return util.error("Missing or invalid Start or End.")
+            if start >= end:
+                return util.error("Start must be before End.")
+
+            try:
+                config = json.loads(request.form["config"])
+                msg = bw_api.set_assignment_config(cid, aid, config)
+
+                if msg:
+                    return util.error("Failed to add assignment to Broadway: {}".format(msg))
+            except json.decoder.JSONDecodeError:
+                return util.error("Failed to decode config JSON")
+
+            visibility = request.form["visibility"] == "visible"
+
+            db.add_assignment(cid, aid, max_runs, quota, start, end, visibility)
+            return "", 204
+
+        @blueprint.route("/staff/course/<cid>/<aid>/edit/", methods=["POST"])
+        @auth.require_auth
+        @auth.require_admin_status
+        def staff_edit_assignment(netid, cid, aid):
+            course = db.get_course(cid)
+            assignment = db.get_assignment(cid, aid)
+            if course is None or assignment is None:
+                return abort(404)
+
+            missing = util.check_missing_fields(request.form, *["max_runs", "quota", "start", "end", "visibility"])
+            if missing:
+                return util.error("Missing fields (%s)." % (", ".join(missing)))
+
+            try:
+                max_runs = int(request.form["max_runs"])
+                if max_runs < 1:
```

Comment:
What if we do not want to allow any pre-deadline runs for an assignment?
Maybe pull this out into a constant MIN_PREDEADLINE_RUNS?